<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>RStan: the R interface to Stan • rstan</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">rstan</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/index.html">Vignettes</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Other Packages
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="http://mc-stan.org/users/interfaces/rstan.html">rstan</a>
    </li>
    <li>
      <a href="http://mc-stan.org/users/interfaces/rstanarm.html">rstanarm</a>
    </li>
    <li>
      <a href="http://mc-stan.org/users/interfaces/bayesplot.html">bayesplot</a>
    </li>
    <li>
      <a href="http://mc-stan.org/users/interfaces/shinystan.html">shinystan</a>
    </li>
    <li>
      <a href="http://mc-stan.org/users/interfaces/rstantools.html">rstantools</a>
    </li>
  </ul>
</li>
<li>
  <a href="http://mc-stan.org">Stan</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="http://discourse.mc-stan.org/">
    <span class="fa fa-users"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/mcmc_stan">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/stan-dev/loo">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>RStan: the R interface to Stan</h1>
                        <h4 class="author">Stan Development Team</h4>
            
            <h4 class="date">2017-07-26</h4>
          </div>

    
    
<div class="contents">
<p>In this vignette we present RStan, the R interface to Stan. Stan is a C++ library for Bayesian inference using the No-U-Turn sampler (a variant of Hamiltonian Monte Carlo) or frequentist inference via optimization. We illustrate the features of RStan through an example in <span class="citation">Gelman et al. (2003)</span>.</p>
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>Stan is a C++ library for Bayesian modeling and inference that primarily uses the No-U-Turn sampler (NUTS) <span class="citation">(Hoffman and Gelman 2012)</span> to obtain posterior simulations given a user-specified model and data. Alternatively, Stan can utilize the LBFGS optimization algorithm to maximize an objective function, such as a log-likelihood. The R package <strong>rstan</strong> provides RStan, the R interface to Stan. The <strong>rstan</strong> package allows one to conveniently fit Stan models from R <span class="citation">(R Core Team 2014)</span> and access the output, including posterior inferences and intermediate quantities such as evaluations of the log posterior density and its gradients.</p>
<p>In this vignette we provide a concise introduction to the functionality included in the <strong>rstan</strong> package. Stan’s website <a href="http://mc-stan.org">mc-stan.org</a> has additional details and provides up-to-date information about how to operate both Stan and its many interfaces including RStan. See, for example, <em>RStan Getting Started</em> <span class="citation">(The Stan Development Team 2014)</span>.</p>
</div>
<div id="prerequisites" class="section level2">
<h2 class="hasAnchor">
<a href="#prerequisites" class="anchor"></a>Prerequisites</h2>
<p>Stan has a modeling language, which is similar to but not identical to that of the Bayesian graphical modeling package BUGS <span class="citation">(Lunn et al. 2000)</span>. A parser translates a model expressed in the Stan language to C++ code, whereupon it is compiled to an executable program and loaded as a Dynamic Shared Object (DSO) in R which can then be called by the user.</p>
<p>A C++ compiler, such as <a href="http://gcc.gnu.org"><code>g++</code></a> or <a href="http://clang.llvm.org"><code>clang++</code></a>, is required for this process. For instructions on installing a C++ compiler for use with RStan see <a href="https://github.com/stan-dev/rstan/wiki/RStan-Getting-Started#prerequisites">RStan-Getting-Started</a>.</p>
<p>The <strong>rstan</strong> package also depends heavily on several other R packages:</p>
<ul>
<li>
<strong>StanHeaders</strong> (Stan C++ headers)</li>
<li>
<strong>BH</strong> (Boost C++ headers)</li>
<li>
<strong>RcppEigen</strong> (Eigen C++ headers)</li>
<li>
<strong>Rcpp</strong> (facilitates using C++ from R)</li>
<li>
<strong>inline</strong> (compiles C++ for use with R)</li>
</ul>
<p>These dependencies should be automatically installed if you install the <strong>rstan</strong> package via one of the conventional mechanisms.</p>
</div>
<div id="typical-workflow" class="section level2">
<h2 class="hasAnchor">
<a href="#typical-workflow" class="anchor"></a>Typical Workflow</h2>
<p>The following is a typical workflow for using Stan via RStan for Bayesian inference.</p>
<ol style="list-style-type: decimal">
<li>Represent a statistical model by writing its log posterior density (up to an normalizing constant that does not depend on the unknown parameters in the model) using the Stan modeling language. We recommend using a separate file with a <code>.stan</code> extension, although it can also be done using a character string within R.</li>
<li>Translate the Stan program to C++ code using the <code>stanc</code> function.</li>
<li>Compile the C++ code to create a DSO (also called a dynamic link library (DLL)) that can be loaded by R.</li>
<li>Run the DSO to sample from the posterior distribution.</li>
<li>Diagnose non-convergence of the MCMC chains.</li>
<li>Conduct inference based on the posterior sample (the MCMC draws from the posterior distribution).</li>
</ol>
<p>Conveniently, steps 2, 3, and 4, above, are all performed implicitly by a single call to the <code>stan</code> function.</p>
</div>
<div id="example" class="section level2">
<h2 class="hasAnchor">
<a href="#example" class="anchor"></a>Example</h2>
<p>Throughout the rest of the vignette we’ll use a hierarchical meta-analysis model described in section 5.5 of <span class="citation">Gelman et al. (2003)</span> as a running example. A hierarchical model is used to model the effect of coaching programs on college admissions tests. The data, shown in the table below, summarize the results of experiments conducted in eight high schools, with an estimated standard error for each. These data and model are of historical interest as an example of full Bayesian inference <span class="citation">(Rubin 1981)</span>. For short, we call this the <em>Eight Schools</em> examples.</p>
<table class="table">
<thead><tr class="header">
<th>School</th>
<th>Estimate (<span class="math inline">\(y_j\)</span>)</th>
<th>Standard Error (<span class="math inline">\(\sigma_j\)</span>)</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>A</td>
<td>28</td>
<td>15</td>
</tr>
<tr class="even">
<td>B</td>
<td>8</td>
<td>10</td>
</tr>
<tr class="odd">
<td>C</td>
<td>-3</td>
<td>16</td>
</tr>
<tr class="even">
<td>D</td>
<td>7</td>
<td>11</td>
</tr>
<tr class="odd">
<td>E</td>
<td>-1</td>
<td>9</td>
</tr>
<tr class="even">
<td>F</td>
<td>1</td>
<td>11</td>
</tr>
<tr class="odd">
<td>G</td>
<td>18</td>
<td>10</td>
</tr>
<tr class="even">
<td>H</td>
<td>12</td>
<td>18</td>
</tr>
</tbody>
</table>
<p>We use the Eight Schools example here because it is simple but also represents a nontrivial Markov chain simulation problem in that there is dependence between the parameters of original interest in the study — the effects of coaching in each of the eight schools — and the hyperparameter representing the variation of these effects in the modeled population. Certain implementations of a Gibbs sampler or a Hamiltonian Monte Carlo sampler can be slow to converge in this example.</p>
<p>The statistical model of interest is specified as</p>
<p><span class="math display">\[
\begin{aligned} 
y_j &amp;\sim \mathsf{Normal}(\theta_j, \sigma_j), \quad j=1,\ldots,8 \\
\theta_j &amp;\sim \mathsf{Normal}(\mu, \tau), \quad j=1,\ldots,8 \\
p(\mu, \tau) &amp;\propto 1,
\end{aligned}
\]</span></p>
<p>where each <span class="math inline">\(\sigma_j\)</span> is assumed known.</p>
<div id="write-a-stan-program" class="section level3">
<h3 class="hasAnchor">
<a href="#write-a-stan-program" class="anchor"></a>Write a Stan Program</h3>
<p>RStan allows a Stan program to be coded in a text file (typically with suffix <code>.stan</code>) or in a R character vector (of length one). We put the following code for the Eight Schools model into the file <code>schools.stan</code>:</p>
<pre><code>data {
  int&lt;lower=0&gt; J;          // number of schools 
  real y[J];               // estimated treatment effects
  real&lt;lower=0&gt; sigma[J];  // s.e. of effect estimates 
}
parameters {
  real mu; 
  real&lt;lower=0&gt; tau;
  vector[J] eta;
}
transformed parameters {
  vector[J] theta;
  theta = mu + tau * eta;
}
model {
  target += normal_lpdf(eta | 0, 1);
  target += normal_lpdf(y | theta, sigma);
}</code></pre>
<p>The first section of the Stan program above, the <code>data</code> block, specifies the data that is conditioned upon in Bayes Rule: the number of schools, <span class="math inline">\(J\)</span>, the vector of estimates, <span class="math inline">\((y_1, \ldots, y_J)\)</span>, and the vector of standard errors of the estimates <span class="math inline">\((\sigma_{1}, \ldots, \sigma_{J})\)</span>. Data are declared as integer or real and can be vectors (or, more generally, arrays) if dimensions are specified. Data can also be constrained; for example, in the above model <span class="math inline">\(J\)</span> has been restricted to be at least <span class="math inline">\(1\)</span> and the components of <span class="math inline">\(\sigma_y\)</span> must all be positive.</p>
<p>The <code>parameters</code> block declares the parameters whose posterior distribution is sought. These are the the mean, <span class="math inline">\(\mu\)</span>, and standard deviation, <span class="math inline">\(\tau\)</span>, of the school effects, plus the <em>standardized</em> school-level effects <span class="math inline">\(\eta\)</span>. In this model, we let the undstandardized school-level effects, <span class="math inline">\(\theta\)</span>, be a transformed parameter constructed by scaling the the standardized effects by <span class="math inline">\(\tau\)</span> and shifting them by <span class="math inline">\(\mu\)</span> rather than directly declaring <span class="math inline">\(\theta\)</span> as a parameter. By parameterizing the model this way, the sampler runs more efficiently because the resulting multivariate geometry is more amendable to Hamiltonian Monte Carlo <span class="citation">(Neal 2011)</span>.</p>
<p>Finally, the <code>model</code> block looks similar to standard statistical notation. (Just be careful: the second argument to Stan’s normal<span class="math inline">\((\cdot,\cdot)\)</span> distribution is the standard deviation, not the variance as is usual in statistical notation.) We have written the model in vector notation, which allows Stan to make use of more efficient algorithmic differentiation (AD). It would also be possible — but less efficient — to write the model by replacing <code>normal_lpdf(y | theta,sigma)</code> with a loop over the <span class="math inline">\(J\)</span> schools,</p>
<pre><code>for (j in 1:J) 
  target += normal_lpdf(y[j] | theta[j],sigma[j]);</code></pre>
<p>Stan has versions of many of the most useful R functions for statistical modeling, including probability distributions, matrix operations, and various special functions. However, the names of the Stan functions may differ from their R counterparts and, more subtly, the parameterizations of probability distributions in Stan may differ from those in R for the same distribution. To mitigate this problem, the <code>lookup</code> function can be passed an R function or character string naming an R function, and RStan will attempt to look up the corresponding Stan function, display its arguments, and give the page number in <span class="citation">The Stan Development Team (2016)</span> where the function is discussed.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/lookup.html">lookup</a></span>(<span class="st">"dnorm"</span>)</code></pre></div>
<pre><code>    StanFunction                         Arguments ReturnType Page
353  normal_lpdf (reals y | reals mu, reals sigma)       real  493
354       normal                                 ~       real  493</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/lookup.html">lookup</a></span>(dwilcox)   <span class="co"># no corresponding Stan function</span></code></pre></div>
<pre><code>[1] "no matching Stan functions"</code></pre>
<p>If the <code>lookup</code> function fails to find an R function that corresponds to a Stan function, it will treat its argument as a regular expression and attempt to find matches with the names of Stan functions.</p>
</div>
<div id="preparing-the-data" class="section level3">
<h3 class="hasAnchor">
<a href="#preparing-the-data" class="anchor"></a>Preparing the Data</h3>
<p>The <code>stan</code> function accepts data as a named list, a character vector of object names, or an <code>environment</code>. Alternatively, the <code>data</code> argument can be omitted and R will search for objects that have the same names as those declared in the <code>data</code> block of the Stan program. Here is the data for the Eight Schools example:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">schools_data &lt;-<span class="st"> </span><span class="kw">list</span>(
  <span class="dt">J =</span> <span class="dv">8</span>,
  <span class="dt">y =</span> <span class="kw">c</span>(<span class="dv">28</span>,  <span class="dv">8</span>, -<span class="dv">3</span>,  <span class="dv">7</span>, -<span class="dv">1</span>,  <span class="dv">1</span>, <span class="dv">18</span>, <span class="dv">12</span>),
  <span class="dt">sigma =</span> <span class="kw">c</span>(<span class="dv">15</span>, <span class="dv">10</span>, <span class="dv">16</span>, <span class="dv">11</span>,  <span class="dv">9</span>, <span class="dv">11</span>, <span class="dv">10</span>, <span class="dv">18</span>)
)</code></pre></div>
<p>It would also be possible (indeed, encouraged) to read in the data from a file rather than to directly enter the numbers in the R script.</p>
</div>
<div id="sample-from-the-posterior-distribution" class="section level3">
<h3 class="hasAnchor">
<a href="#sample-from-the-posterior-distribution" class="anchor"></a>Sample from the Posterior Distribution</h3>
<p>Next, we can call the <code>stan</code> function to draw posterior samples:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(rstan)
fit1 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/stan.html">stan</a></span>(
  <span class="dt">file =</span> <span class="st">"schools.stan"</span>,  <span class="co"># Stan program</span>
  <span class="dt">data =</span> schools_data,    <span class="co"># named list of data</span>
  <span class="dt">chains =</span> <span class="dv">4</span>,             <span class="co"># number of Markov chains</span>
  <span class="dt">warmup =</span> <span class="dv">1000</span>,          <span class="co"># number of warmup iterations per chain</span>
  <span class="dt">iter =</span> <span class="dv">2000</span>,            <span class="co"># total number of iterations per chain</span>
  <span class="dt">cores =</span> <span class="dv">2</span>,              <span class="co"># number of cores (using 2 just for the vignette)</span>
  <span class="dt">refresh =</span> <span class="dv">1000</span>          <span class="co"># show progress every 'refresh' iterations</span>
  )</code></pre></div>
<pre><code>Warning: There were 2 divergent transitions after warmup. Increasing adapt_delta above 0.8 may help. See
http://mc-stan.org/misc/warnings.html#divergent-transitions-after-warmup</code></pre>
<pre><code>Warning: Examine the pairs() plot to diagnose sampling problems</code></pre>
<p>The <code>stan</code> function wraps the following three steps:</p>
<ul>
<li>Translate a model in Stan code to C++ code</li>
<li>Compile the C++ code to a dynamic shared object (DSO) and load the DSO</li>
<li>Sample given some user-specified data and other settings</li>
</ul>
<p>A single call to <code>stan</code> performs all three steps, but they can also be executed one by one (see the help pages for <code>stanc</code>, <code>stan_model</code>, and <code>sampling</code>), which can be useful for debugging. In addition, Stan saves the DSO so that when the same model is fit again (possibly with new data and settings) we can avoid recompilation. If an error happens after the model is compiled but before sampling (e.g., problems with inputs like data and initial values), we can still reuse the compiled model.</p>
<p>The <code>stan</code> function returns a stanfit object, which is an S4 object of class <code>"stanfit"</code>. For those who are not familiar with the concept of class and S4 class in R, refer to <span class="citation">Chambers (2008)</span>. An S4 class consists of some attributes (data) to model an object and some methods to model the behavior of the object. From a user’s perspective, once a stanfit object is created, we are mainly concerned about what methods are defined.</p>
<p>If no error occurs, the returned stanfit object includes the sample drawn from the posterior distribution for the model parameters and other quantities defined in the model. If there is an error (e.g. a syntax error in the Stan program), <code>stan</code> will either quit or return a stanfit object that contains no posterior draws.</p>
<p>For class <code>"stanfit"</code>, many methods such as <code>print</code> and <code>plot</code> are defined for working with the MCMC sample. For example, the following shows a summary of the parameters from the Eight Schools model using the <code>print</code> method:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(fit1, <span class="dt">pars=</span><span class="kw">c</span>(<span class="st">"theta"</span>, <span class="st">"mu"</span>, <span class="st">"tau"</span>, <span class="st">"lp__"</span>), <span class="dt">probs=</span><span class="kw">c</span>(.<span class="dv">1</span>,.<span class="dv">5</span>,.<span class="dv">9</span>))</code></pre></div>
<pre><code>Inference for Stan model: schools.
4 chains, each with iter=2000; warmup=1000; thin=1; 
post-warmup draws per chain=1000, total post-warmup draws=4000.

           mean se_mean   sd    10%    50%    90% n_eff Rhat
theta[1]  11.33    0.16 8.41   2.30  10.01  22.29  2924    1
theta[2]   7.92    0.10 6.34   0.39   7.82  15.73  4000    1
theta[3]   6.21    0.12 7.89  -3.34   6.77  15.28  4000    1
theta[4]   7.74    0.10 6.58  -0.43   7.80  15.97  4000    1
theta[5]   5.07    0.10 6.37  -3.25   5.63  12.54  4000    1
theta[6]   6.17    0.10 6.56  -2.20   6.47  14.00  4000    1
theta[7]  10.64    0.11 6.67   2.80   9.98  19.27  4000    1
theta[8]   8.52    0.13 7.96  -0.71   8.33  17.92  4000    1
mu         7.94    0.10 5.00   1.55   7.98  14.32  2442    1
tau        6.43    0.13 5.53   0.94   5.09  13.36  1721    1
lp__     -39.66    0.07 2.68 -43.18 -39.40 -36.47  1380    1

Samples were drawn using NUTS(diag_e) at Wed Jul 26 21:53:10 2017.
For each parameter, n_eff is a crude measure of effective sample size,
and Rhat is the potential scale reduction factor on split chains (at 
convergence, Rhat=1).</code></pre>
<p>The last line of this output, <code>lp__</code>, is the logarithm of the (unnormalized) posterior density as calculated by Stan. This log density can be used in various ways for model evaluation and comparison (see, e.g., <span class="citation">Vehtari and Ojanen (2012)</span>).</p>
<div id="arguments-to-the-stan-function" class="section level4">
<h4 class="hasAnchor">
<a href="#arguments-to-the-stan-function" class="anchor"></a>Arguments to the <code>stan</code> Function</h4>
<p>The primary arguments for sampling (in functions <code>stan</code> and <code>sampling</code>) include data, initial values, and the options of the sampler such as <code>chains</code>, <code>iter</code>, and <code>warmup</code>. In particular, <code>warmup</code> specifies the number of iterations that are used by the NUTS sampler for the adaptation phase before sampling begins. After the warmup, the sampler turns off adaptation and continues until a total of <code>iter</code> iterations (including <code>warmup</code>) have been completed. There is no theoretical guarantee that the draws obtained during warmup are from the posterior distribution, so the warmup draws should only be used for diagnosis and not inference. The summaries for the parameters shown by the <code>print</code> method are calculated using only post-warmup draws.</p>
<p>The optional <code>init</code> argument can be used to specify initial values for the Markov chains. There are several ways to specify initial values, and the details can be found in the documentation of the <code>stan</code> function. The vast majority of the time it is adequate to allow Stan to generate its own initial values randomly. However, sometimes it is better to specify the initial values for at least a subset of the objects declared in the <code>parameters</code> block of a Stan program.</p>
<p>Stan uses a random number generator (RNG) that supports parallelism. The initialization of the RNG is determined by the arguments <code>seed</code> and <code>chain_id</code>. Even if we are running multiple chains from one call to the <code>stan</code> function we only need to specify one seed, which is randomly generated by R if not specified.</p>
</div>
<div id="data-preprocessing-and-passing" class="section level4">
<h4 class="hasAnchor">
<a href="#data-preprocessing-and-passing" class="anchor"></a>Data Preprocessing and Passing</h4>
<p>The data passed to <code>stan</code> will go through a preprocessing procedure. The details of this preprocessing are documented in the documentation for the <code>stan</code> function. Here we stress a few important steps. First, RStan allows the user to pass more objects as data than what is declared in the <code>data</code> block (silently omitting any unnecessary objects). In general, an element in list of data passed to Stan from R should be numeric and its dimension should match the declaration in the <code>data</code> block of the model. So for example, the <code>factor</code> type in R is not supported as a data element for RStan and must be converted to integer codes via <code>as.integer</code>. The Stan modeling language distinguishes between integers and doubles (type <code>int</code> and <code>real</code> in Stan modeling language, respectively). The <code>stan</code> function will convert some R data (which is double-precision usually) to integers if possible.</p>
<p>The Stan language has scalars and other types that are sets of scalars, e.g. vectors, matrices, and arrays. As R does not have true scalars, RStan treats vectors of length one as scalars. However, consider a model with a <code>data</code> block defined as</p>
<pre><code>data {                
  int&lt;lower=1&gt; N;      
  real y[N];
} </code></pre>
<p>in which <code>N</code> can be <span class="math inline">\(1\)</span> as a special case. So if we know that <code>N</code> is always larger than <span class="math inline">\(1\)</span>, we can use a vector of length <code>N</code> in R as the data input for <code>y</code> (for example, a vector created by <code>y &lt;- rnorm(N)</code>). If we want to prevent RStan from treating the input data for <code>y</code> as a scalar when <span class="math inline">\(N`\)</span> is <span class="math inline">\(1\)</span>, we need to explicitly make it an array as the following R code shows:</p>
<pre><code>y &lt;- as.array(y)</code></pre>
<p>Stan cannot handle missing values in data automatically, so no element of the data can contain <code>NA</code> values. An important step in RStan’s data preprocessing is to check missing values and issue an error if any are found. There are, however, various ways of writing Stan programs that account for missing data (see <span class="citation">The Stan Development Team (2016)</span>).</p>
</div>
</div>
<div id="methods-for-the-stanfit-class" class="section level3">
<h3 class="hasAnchor">
<a href="#methods-for-the-stanfit-class" class="anchor"></a>Methods for the <code>"stanfit"</code> Class</h3>
<p>The other vignette included with the <strong>rstan</strong> package discusses stanfit objects in greater detail and gives examples of accessing the most important content contained in the objects (e.g., posterior draws, diagnostic summaries). Also, a full list of available methods can be found in the documentation for the <code>"stanfit"</code> class at <code>help("stanfit", "rstan")</code>. Here we give only a few examples.</p>
<p>The <code>plot</code> method for stanfit objects provides various graphical overviews of the output. The default plot shows posterior uncertainty intervals (by default 80% (inner) and 95% (outer)) and the posterior median for all the parameters as well as <code>lp__</code> (the log of posterior density function up to an additive constant):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(fit1)</code></pre></div>
<pre><code>'pars' not specified. Showing first 10 parameters by default.</code></pre>
<pre><code>ci_level: 0.8 (80% intervals)</code></pre>
<pre><code>outer_level: 0.95 (95% intervals)</code></pre>
<p><img src="rstan_files/figure-html/stanfit-plot-1.png" width="672" style="display: block; margin: auto;"></p>
<p>The optional <code>plotfun</code> argument can be used to select among the various available plots. See <code>help("plot,stanfit-method")</code>.</p>
<p>The <code>traceplot</code> method is used to plot the time series of the posterior draws. If we include the warmup draws by setting <code>inc_warmup=TRUE</code>, the background color of the warmup area is different from the post-warmup phase:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/stanfit-method-traceplot.html">traceplot</a></span>(fit1, <span class="dt">pars =</span> <span class="kw">c</span>(<span class="st">"mu"</span>, <span class="st">"tau"</span>), <span class="dt">inc_warmup =</span> <span class="ot">TRUE</span>, <span class="dt">nrow =</span> <span class="dv">2</span>)</code></pre></div>
<p><img src="rstan_files/figure-html/stanfit-traceplot-1.png" width="672" style="display: block; margin: auto;"></p>
<p>To assess the convergence of the Markov chains, in addition to visually inspecting traceplots we can calculate the split <span class="math inline">\(\hat{R}\)</span> statistic. Split <span class="math inline">\(\hat{R}\)</span> is an updated version of the <span class="math inline">\(\hat{R}\)</span> statistic proposed in <span class="citation">Gelman and Rubin (1992)</span> that is based on splitting each chain into two halves. See the Stan manual for more details. The estimated <span class="math inline">\(\hat{R}\)</span> for each parameter is included as one of the columns in the output from the <code>summary</code> and <code>print</code> methods.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(fit1, <span class="dt">pars =</span> <span class="kw">c</span>(<span class="st">"mu"</span>, <span class="st">"tau"</span>))</code></pre></div>
<pre><code>Inference for Stan model: schools.
4 chains, each with iter=2000; warmup=1000; thin=1; 
post-warmup draws per chain=1000, total post-warmup draws=4000.

    mean se_mean   sd  2.5%  25%  50%   75% 97.5% n_eff Rhat
mu  7.94    0.10 5.00 -1.75 4.63 7.98 11.10 17.82  2442    1
tau 6.43    0.13 5.53  0.24 2.35 5.09  9.04 20.45  1721    1

Samples were drawn using NUTS(diag_e) at Wed Jul 26 21:53:10 2017.
For each parameter, n_eff is a crude measure of effective sample size,
and Rhat is the potential scale reduction factor on split chains (at 
convergence, Rhat=1).</code></pre>
<p>Again, see the additional vignette on stanfit objects for more details.</p>
</div>
<div id="sampling-difficulties" class="section level3">
<h3 class="hasAnchor">
<a href="#sampling-difficulties" class="anchor"></a>Sampling Difficulties</h3>
<p>The best way to visualize the output of a model is through the ShinyStan interface, which can be accessed via the <a href="https://cran.r-project.org/package=shinystan"><strong>shinystan</strong></a> R package. ShinyStan facilitates both the visualization of parameter distributions and diagnosing problems with the sampler. The documentation for the <strong>shinystan</strong> package provides instructions for using the interface with stanfit objects.</p>
<p>In addition to using ShinyStan, it is also possible to diagnose some sampling problems using functions in the <strong>rstan</strong> package. The <code>get_sampler_params</code> function returns information on parameters related the performance of the sampler:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># all chains combined</span>
sampler_params &lt;-<span class="st"> </span><span class="kw"><a href="../reference/stanfit-class.html">get_sampler_params</a></span>(fit1, <span class="dt">inc_warmup =</span> <span class="ot">TRUE</span>)
<span class="kw">summary</span>(<span class="kw">do.call</span>(rbind, sampler_params), <span class="dt">digits =</span> <span class="dv">2</span>)</code></pre></div>
<pre><code> accept_stat__    stepsize__      treedepth__   n_leapfrog__
 Min.   :0.00   Min.   : 0.022   Min.   :0.0   Min.   :  1  
 1st Qu.:0.78   1st Qu.: 0.297   1st Qu.:3.0   1st Qu.:  7  
 Median :0.95   Median : 0.375   Median :3.0   Median : 15  
 Mean   :0.83   Mean   : 0.396   Mean   :3.4   Mean   : 13  
 3rd Qu.:0.99   3rd Qu.: 0.389   3rd Qu.:4.0   3rd Qu.: 15  
 Max.   :1.00   Max.   :14.386   Max.   :7.0   Max.   :127  
  divergent__        energy__ 
 Min.   :0.0000   Min.   :35  
 1st Qu.:0.0000   1st Qu.:42  
 Median :0.0000   Median :44  
 Mean   :0.0097   Mean   :45  
 3rd Qu.:0.0000   3rd Qu.:47  
 Max.   :1.0000   Max.   :62  </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># each chain separately</span>
<span class="kw">lapply</span>(sampler_params, summary, <span class="dt">digits =</span> <span class="dv">2</span>)</code></pre></div>
<pre><code>[[1]]
 accept_stat__    stepsize__      treedepth__   n_leapfrog__
 Min.   :0.00   Min.   : 0.034   Min.   :0.0   Min.   :  1  
 1st Qu.:0.80   1st Qu.: 0.331   1st Qu.:3.0   1st Qu.:  7  
 Median :0.95   Median : 0.331   Median :4.0   Median : 15  
 Mean   :0.84   Mean   : 0.394   Mean   :3.5   Mean   : 13  
 3rd Qu.:0.99   3rd Qu.: 0.394   3rd Qu.:4.0   3rd Qu.: 15  
 Max.   :1.00   Max.   :14.386   Max.   :7.0   Max.   :127  
  divergent__       energy__ 
 Min.   :0.000   Min.   :36  
 1st Qu.:0.000   1st Qu.:42  
 Median :0.000   Median :44  
 Mean   :0.011   Mean   :45  
 3rd Qu.:0.000   3rd Qu.:47  
 Max.   :1.000   Max.   :58  

[[2]]
 accept_stat__    stepsize__      treedepth__   n_leapfrog__
 Min.   :0.00   Min.   : 0.022   Min.   :0.0   Min.   :  1  
 1st Qu.:0.78   1st Qu.: 0.377   1st Qu.:3.0   1st Qu.:  7  
 Median :0.94   Median : 0.377   Median :3.0   Median :  7  
 Mean   :0.82   Mean   : 0.432   Mean   :3.3   Mean   : 12  
 3rd Qu.:0.99   3rd Qu.: 0.423   3rd Qu.:4.0   3rd Qu.: 15  
 Max.   :1.00   Max.   :13.722   Max.   :7.0   Max.   :127  
  divergent__        energy__ 
 Min.   :0.0000   Min.   :36  
 1st Qu.:0.0000   1st Qu.:42  
 Median :0.0000   Median :44  
 Mean   :0.0095   Mean   :45  
 3rd Qu.:0.0000   3rd Qu.:47  
 Max.   :1.0000   Max.   :62  

[[3]]
 accept_stat__    stepsize__     treedepth__   n_leapfrog__
 Min.   :0.00   Min.   :0.042   Min.   :0.0   Min.   : 1   
 1st Qu.:0.71   1st Qu.:0.359   1st Qu.:3.0   1st Qu.: 7   
 Median :0.93   Median :0.375   Median :3.0   Median :15   
 Mean   :0.81   Mean   :0.392   Mean   :3.3   Mean   :13   
 3rd Qu.:0.98   3rd Qu.:0.375   3rd Qu.:4.0   3rd Qu.:15   
 Max.   :1.00   Max.   :7.434   Max.   :6.0   Max.   :63   
  divergent__       energy__ 
 Min.   :0.000   Min.   :35  
 1st Qu.:0.000   1st Qu.:42  
 Median :0.000   Median :44  
 Mean   :0.012   Mean   :45  
 3rd Qu.:0.000   3rd Qu.:47  
 Max.   :1.000   Max.   :59  

[[4]]
 accept_stat__    stepsize__     treedepth__   n_leapfrog__
 Min.   :0.00   Min.   :0.044   Min.   :0.0   Min.   : 1   
 1st Qu.:0.83   1st Qu.:0.297   1st Qu.:3.0   1st Qu.: 7   
 Median :0.96   Median :0.297   Median :4.0   Median :15   
 Mean   :0.85   Mean   :0.367   Mean   :3.5   Mean   :14   
 3rd Qu.:0.99   3rd Qu.:0.391   3rd Qu.:4.0   3rd Qu.:15   
 Max.   :1.00   Max.   :5.127   Max.   :6.0   Max.   :63   
  divergent__        energy__ 
 Min.   :0.0000   Min.   :36  
 1st Qu.:0.0000   1st Qu.:42  
 Median :0.0000   Median :44  
 Mean   :0.0065   Mean   :45  
 3rd Qu.:0.0000   3rd Qu.:47  
 Max.   :1.0000   Max.   :60  </code></pre>
<p>Here we see that there are a small number of divergent transitions, which are identified by <code>divergent__</code> being <span class="math inline">\(1\)</span>. Ideally, there should be no divergent transitions after the warmup phase. The best way to try to eliminate divergent transitions is by increasing the target acceptance probability, which by default is <span class="math inline">\(0.8\)</span>. In this case the mean of <code>accept_stat__</code> is close to <span class="math inline">\(0.8\)</span> for all chains, but has a very skewed distribution because the median is near <span class="math inline">\(0.95\)</span>. We could go back and call <code>stan</code> again and specify the optional argument <code>control=list(adapt_delta=0.9)</code> to try to eliminate the divergent transitions. However, sometimes when the target acceptance rate is high, the stepsize is very small and the sampler hits its limit on the number of leapfrog steps it can take per iteration. In this case, it is a non-issue because each chain has a <code>treedepth__</code> of at most <span class="math inline">\(7\)</span> and the default is <span class="math inline">\(10\)</span>. But if any <code>treedepth__</code> were <span class="math inline">\(11\)</span>, then it would be wise to increase the limit by passing <code>control=list(max_treedepth=12)</code> (for example) to <code>stan</code>. See the vignette on stanfit objects for more on the structure of the object returned by <code>get_sampler_params</code>.</p>
<p>We can also make a graphical representation of (much of the) the same information using <code>pairs</code>. The “pairs”" plot can be used to get a sense of whether any sampling difficulties are occurring in the tails or near the mode:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">pairs</span>(fit1, <span class="dt">pars =</span> <span class="kw">c</span>(<span class="st">"mu"</span>, <span class="st">"tau"</span>, <span class="st">"lp__"</span>), <span class="dt">las =</span> <span class="dv">1</span>)</code></pre></div>
<p><img src="rstan_files/figure-html/pairs-plot-1.png" width="672" style="display: block; margin: auto;"></p>
<p>In the plot above, the marginal distribution of each selected parameter is included as a histogram along the diagnoal. By default, draws with below-median <code>accept_stat__</code> (MCMC proposal acceptance rate) are plotted below the diagonal and those with above-median <code>accept_stat__</code> are plotted above the diagonal (this can be changed using the <code>condition</code> argument). Each off-diagonal square represents a bivariate distribution of the draws for the intersection of the row-variable and the column-variable. Ideally, the below-diagonal intersection and the above-diagonal intersection of the same two variables should have distributions that are mirror images of each other. Any yellow points would indicate transitions where the maximum <code>treedepth__</code> was hit, and red points indicate a divergent transition.</p>
</div>
</div>
<div id="additional-topics" class="section level2">
<h2 class="hasAnchor">
<a href="#additional-topics" class="anchor"></a>Additional Topics</h2>
<div id="user-defined-stan-functions" class="section level3">
<h3 class="hasAnchor">
<a href="#user-defined-stan-functions" class="anchor"></a>User-defined Stan Functions</h3>
<p>Stan also permits users to define their own functions that can be used throughout a Stan program. These functions are defined in the <code>functions</code> block. The <code>functions</code> block is optional but, if it exists, it must come before any other block. This mechanism allows users to implement statistical distributions or other functionality that is not currently available in Stan. However, even if the user’s function merely wraps calls to existing Stan functions, the code in the <code>model</code> block can be much more readible if several lines of Stan code that accomplish one (or perhaps two) task(s) are replaced by a call to a user-defined function.</p>
<p>Another reason to utilize user-defined functions is that RStan provides the <code>expose_stan_functions</code> function for exporting such functions to the R global environment so that they can be tested in R to ensure they are working properly. For example,</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">model_code &lt;-
<span class="st">'</span>
<span class="st">functions {</span>
<span class="st">  real standard_normal_rng() {</span>
<span class="st">    return normal_rng(0,1);</span>
<span class="st">  }</span>
<span class="st">}</span>
<span class="st">model {}</span>
<span class="st">'</span>
<span class="kw"><a href="../reference/expose_stan_functions.html">expose_stan_functions</a></span>(<span class="kw"><a href="../reference/stanc.html">stanc</a></span>(<span class="dt">model_code =</span> model_code))
<span class="kw">standard_normal_rng</span>(<span class="dt">seed =</span> <span class="dv">1</span>)</code></pre></div>
<pre><code>[1] -0.9529876</code></pre>
</div>
<div id="the-log-posterior-function-and-gradient" class="section level3">
<h3 class="hasAnchor">
<a href="#the-log-posterior-function-and-gradient" class="anchor"></a>The Log-Posterior (function and gradient)</h3>
<p>Stan defines the log of the probability density function of a posterior distribution up to an unknown additive constant. We use <code>lp__</code> to represent the realizations of this log kernel at each iteration (and <code>lp__</code> is treated as an unknown in the summary and the calculation of split <span class="math inline">\(\hat{R}\)</span> and effective sample size).</p>
<p>A nice feature of the <strong>rstan</strong> package is that it exposes functions for calculating both <code>lp__</code> and its gradients for a given stanfit object. These two functions are <code>log_prob</code> and <code>grad_log_prob</code>, respectively. Both take parameters on the <em>unconstrained</em> space, even if the support of a parameter is not the whole real line. The Stan manual <span class="citation">(The Stan Development Team 2016)</span> has full details on the particular transformations Stan uses to map from the entire real line to some subspace of it (and vice-versa).</p>
<p>It maybe the case that the number of unconstrained parameters might be less than the total number of parameters. For example, for a simplex parameter of length <span class="math inline">\(K\)</span>, there are actually only <span class="math inline">\(K-1\)</span> unconstrained parameters because of the constraint that all elements of a simplex must be nonnegative and sum to one. The <code>get_num_upars</code> method is provided to get the number of unconstrained parameters, while the <code>unconstrain_pars</code> and <code>constrain_pars</code> methods can be used to compute unconstrained and constrained values of parameters respectively. The former takes a list of parameters as input and transforms it to an unconstrained vector, and the latter does the opposite. Using these functions, we can implement other algorithms such as maximum a posteriori estimation of Bayesian models.</p>
</div>
<div id="optimization-in-stan" class="section level3">
<h3 class="hasAnchor">
<a href="#optimization-in-stan" class="anchor"></a>Optimization in Stan</h3>
<p>RStan also provides an interface to Stan’s optimizers, which can be used to obtain a point estimate by maximizing the (perhaps penalized) likelihood function defined by a Stan program. We illustrate this feature using a very simple example: estimating the mean from samples assumed to be drawn from a normal distribution with known standard deviation. That is, we assume</p>
<p><span class="math display">\[y_n \sim \mathsf{Normal}(\mu,1), \quad n = 1, \ldots, N. \]</span></p>
<p>By specifying a prior <span class="math inline">\(p(\mu) \propto 1\)</span>, the maximum a posteriori estimator for <span class="math inline">\(\mu\)</span> is just the sample mean. We don’t need to explicitly code this prior for <span class="math inline">\(\mu\)</span>, as <span class="math inline">\(p(\mu) \propto 1\)</span> is the default if no prior is specified.</p>
<p>We first create an object of class <code>"stanmodel"</code> and then use the <code>optimizing</code> method, to which data and other arguments can be fed.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ocode &lt;-<span class="st"> "</span>
<span class="st">  data {</span>
<span class="st">    int&lt;lower=1&gt; N;</span>
<span class="st">    real y[N];</span>
<span class="st">  } </span>
<span class="st">  parameters {</span>
<span class="st">    real mu;</span>
<span class="st">  } </span>
<span class="st">  model {</span>
<span class="st">    target += normal_lpdf(y | mu, 1);</span>
<span class="st">  } </span>
<span class="st">"</span>

sm &lt;-<span class="st"> </span><span class="kw"><a href="../reference/stan_model.html">stan_model</a></span>(<span class="dt">model_code =</span> ocode)
y2 &lt;-<span class="st"> </span><span class="kw">rnorm</span>(<span class="dv">20</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">mean</span>(y2)</code></pre></div>
<pre><code>[1] -0.2390763</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/stanmodel-method-optimizing.html">optimizing</a></span>(sm, <span class="dt">data =</span> <span class="kw">list</span>(<span class="dt">y =</span> y2, <span class="dt">N =</span> <span class="kw">length</span>(y2)), <span class="dt">hessian =</span> <span class="ot">TRUE</span>)</code></pre></div>
<pre><code>Initial log joint probability = -24.0333
Optimization terminated normally: 
  Convergence detected: gradient norm is below tolerance</code></pre>
<pre><code>$par
        mu 
-0.2390763 

$value
[1] -22.11149

$return_code
[1] 0

$hessian
    mu
mu -20</code></pre>
</div>
<div id="model-compilation" class="section level3">
<h3 class="hasAnchor">
<a href="#model-compilation" class="anchor"></a>Model Compilation</h3>
<p>As mentioned earlier in the vignette, Stan programs are written in the Stan modeling language, translated to C++ code, and then compiled to a dynamic shared object (DSO). The DSO is then loaded by R and executed to draw the posterior sample. The process of compiling C++ code to DSO sometimes takes a while. When the model is the same, we can reuse the DSO from a previous run. The <code>stan</code> function accepts the optional argument <code>fit</code>, which can be used to pass an existing fitted model object so that the compiled model is reused. When reusing a previous fitted model, we can still specify different values for the other arguments to <code>stan</code>, including passing different data to the <code>data</code> argument.</p>
<p>In addition, if fitted models are saved using functions like <code>save</code> and <code>save.image</code>, RStan is able to save DSOs, so that they can be used across R sessions. To avoid saving the DSO, specify <code>save_dso=FALSE</code> when calling the <code>stan</code> function.</p>
<p>If the user executes <code><a href="../reference/rstan_options.html">rstan_options(auto_write = TRUE)</a></code>, then a serialized version of the compiled model will be automatically saved to the hard disk in the same directory as the <code>.stan</code> file or in R’s temporary directory if the Stan program is expressed as a character string. Although this option is not enabled by default due to CRAN policy, it should ordinarily be specified by users in order to eliminate redundant compilation.</p>
<p>Stan runs much faster when the code is compiled at the maximum level of optimization, which is <code>-O3</code> on most C++ compilers. However, the default value is <code>-O2</code> in R, which is appropriate for most R packages but entails a slight slowdown for Stan. You can change this default locally by following the instructions at <a href="https://cran.r-project.org/doc/manuals/r-release/R-admin.html#Customizing-package-compilation">CRAN - Customizing-package-compilation</a>. However, you should be advised that setting <code>CXXFLAGS = -O3</code> may cause adverse side effects for other R packages.</p>
<p>See the documentation for the <code>stanc</code> and <code>stan_model</code> functions for more details on the parsing and compilation of Stan programs.</p>
</div>
<div id="running-multiple-chains-in-parallel" class="section level3">
<h3 class="hasAnchor">
<a href="#running-multiple-chains-in-parallel" class="anchor"></a>Running Multiple Chains in Parallel</h3>
<p>The number of Markov chains to run can be specified using the <code>chains</code> argument to the <code>stan</code> or <code>sampling</code> functions. By default, the chains are executed serially (i.e., one at a time) using the parent R process. There is also an optional <code>cores</code> argument that can be set to the number of chains (if the hardware has sufficient processors and RAM), which is appropriate on most laptops. We typically recommend first calling <code>options(mc.cores=parallel::detectCores())</code> once per R session so that all available cores can be used without needing to manually specify the <code>cores</code> argument.</p>
<p>For users working with a different parallelization scheme (perhaps with a remote cluster), the <strong>rstan</strong> package provides a function called <code>sflist2stanfit</code> for consolidating a list of multiple stanfit objects (created from the same Stan program and using the same number of warmup and sampling iterations) into a single stanfit object. It is important to specify the same seed for all the chains and equally important to use a different chain ID (argument <code>chain_id</code>), the combination of which ensures that the random numbers generated in Stan for all chains are essentially independent. This is handled automatically (internally) when <span class="math inline">\(`cores` &gt; 1\)</span>.</p>
</div>
</div>
<div id="working-with-cmdstan" class="section level2">
<h2 class="hasAnchor">
<a href="#working-with-cmdstan" class="anchor"></a>Working with CmdStan</h2>
<p>The <strong>rstan</strong> package provides some functions for creating data for and reading output from CmdStan, the command line interface to Stan.</p>
<p>First, when Stan reads data or initial values, it supports a subset of the syntax of R dump data formats. So if we use the <code>dump</code> function in base R to prepare data, Stan might not be able to read the contents. The <code>stan_rdump</code> function in <strong>rstan</strong> is designed to dump the data from R to a format that is supported by Stan, with symantics that are very similar to the <code>dump</code> function.</p>
<p>Second, the <code>read_stan_csv</code> function creates a stanfit object from reading the CSV files generated by CmdStan. The resulting stanfit object is compatible with the various methods for diagnostics and posterior analysis.</p>
</div>
<div id="see-also" class="section level2">
<h2 class="hasAnchor">
<a href="#see-also" class="anchor"></a>See Also</h2>
<ul>
<li>The other vignette for the <strong>rstan</strong> package, which shows how to access the contents of stanfit objects.</li>
<li>The very thorough <a href="http://mc-stan.org/documentation/">Stan manual</a> <span class="citation">(The Stan Development Team 2016)</span>.</li>
<li>The <code>stan_demo</code> function, which can be used to fit many of the example models in the manual.</li>
<li>The <a href="https://cran.r-project.org/package=shinystan"><strong>shinystan</strong></a> R package, which provides the ShinyStan GUI for exploring MCMC output.</li>
<li>The <a href="https://cran.r-project.org/package=loo"><strong>loo</strong></a> R package, which is very useful for model comparison using stanfit objects.</li>
<li>The <a href="https://cran.r-project.org/package=rstanarm"><strong>rstanarm</strong></a> R package, which provides a <code>glmer</code>-style interface to Stan.</li>
</ul>
<div id="refs" class="references">
<div id="ref-chambers2010software">
<p>Chambers, John M. 2008. <em>Software for Data Analysis : Programming with R</em>. New York: Springer.</p>
</div>
<div id="ref-GelmanRubin:1992">
<p>Gelman, Andrew, and Donald B. Rubin. 1992. “Inference from Iterative Simulation Using Multiple Sequences.” <em>Statistical Science</em> 7 (4): 457–72.</p>
</div>
<div id="ref-GelmanCarlinSternRubin:2003">
<p>Gelman, Andrew, J. B. Carlin, Hal S. Stern, and Donald B. Rubin. 2003. <em>Bayesian Data Analysis</em>. 2nd ed. London: CRC Press.</p>
</div>
<div id="ref-hoffman-gelman:2012">
<p>Hoffman, Matthew D., and Andrew Gelman. 2012. “The No-U-Turn Sampler: Adaptively Setting Path Lengths in Hamiltonian Monte Carlo.” <em>Journal of Machine Learning Research</em>.</p>
</div>
<div id="ref-WinBUGS">
<p>Lunn, D.J., A. Thomas, N. Best, and D. Spiegelhalter. 2000. “WinBUGS — a Bayesian Modelling Framework: Concepts, Structure, and Extensibility.” <em>Statistics and Computing</em>, 325–37.</p>
</div>
<div id="ref-Neal:2011">
<p>Neal, Radford. 2011. “MCMC Using Hamiltonian Dynamics.” In <em>Handbook of Markov Chain Monte Carlo</em>, edited by Steve Brooks, Andrew Gelman, Galin L. Jones, and Xiao-Li Meng, 116–62. Chapman; Hall/CRC.</p>
</div>
<div id="ref-rprj">
<p>R Core Team. 2014. <em>R: A Language and Environment for Statistical Computing</em>. Vienna, Austria: R Foundation for Statistical Computing. <a href="https://www.R-project.org/" class="uri">https://www.R-project.org/</a>.</p>
</div>
<div id="ref-Rubin1981">
<p>Rubin, Donald B. 1981. “Estimation in Parallel Randomized Experiments.” <em>Journal of Educational and Behavioral Statistics</em> 6 (4): 377–401.</p>
</div>
<div id="ref-rstangettingstarted2012">
<p>The Stan Development Team. 2014. “RStan Getting Started.” <a href="http://mc-stan.org/" class="uri">http://mc-stan.org/</a>.</p>
</div>
<div id="ref-StanManual">
<p>———. 2016. <em>Stan Modeling Language: User’s Guide and Reference Manual</em>. <a href="http://mc-stan.org" class="uri">http://mc-stan.org</a>.</p>
</div>
<div id="ref-Vehtari2012">
<p>Vehtari, A., and J. Ojanen. 2012. “A Survey of Bayesian Predictive Methods for Model Assessment, Selection and Comparison.” <em>Statistics Surveys</em> 6: 142–228.</p>
</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#prerequisites">Prerequisites</a></li>
      <li><a href="#typical-workflow">Typical Workflow</a></li>
      <li><a href="#example">Example</a></li>
      <li><a href="#additional-topics">Additional Topics</a></li>
      <li><a href="#working-with-cmdstan">Working with CmdStan</a></li>
      <li><a href="#see-also">See Also</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Jiqiang Guo, Jonah Gabry, Ben Goodrich.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
